File: Arrays/MajorlyElement2/main.py
class MajorlyElement2:
    def majorElement(self,nums):
        n = len(nums)
        l1 = float('-inf')
        l2 = float('-inf')
        m1 = 0
        m2 = 0
        c1 = 0
        c2 = 0
        ans = []
        for i in range(n):
            if nums[i] != l2 and m1==0:
                l1 = nums[i]
                m1 = 1
            elif nums[i] != l1 and m2==0:
                l2 = nums[i]
                m2 = 1
            elif nums[i] == l1:
                m1 += 1
            elif nums[i] == l2:
                m2 += 1
            else :
                m1 -= 1
                m2 -= 1 
            
            
        for i in range(n):
            if nums[i] == l1:
                c1 += 1
            elif nums[i] == l2:
                c2 += 1

        if c1 > n // 3:
            ans.append(l1)
        if c2 > n // 3:
            ans.append(l2)

        return ans 
    
    
n = int(input("Enter the number of elements in the array: "))
nums = list(map(int, input("Enter the elements of the array: ").split()))
obj = MajorlyElement2()
result = obj.majorElement(nums)
print("The majority elements are:", result)

File: Arrays/MajorlyElement2/main.cpp
#include<bits/stdc++.h>
using namespace std;

class MajorityElement2{
public:
    vector<int>majorElement(vector<int>&nums){
        int n = nums.size();
        vector<int>ans;
        int leader1=INT_MIN,leader2=INT_MIN;
        int maj1 = 0,maj2 = 0;
        for(int i=0;i<n;i++){
            if(nums[i] != leader2 && maj1==0){
                leader1 = nums[i];
                maj1=1;
            }
            else if(nums[i]!=leader1 && maj2==0){
                leader2 = nums[i];
                maj2=1;
            }
            else if(nums[i]==leader1){
                maj1++;
            }
            else if(nums[i]==leader2){
                maj2++;
            }
            else{
                maj1--;
                maj2--;
            }
        }
        int count1=0,count2=0;
        for(int i=0;i<n;i++){
            if(nums[i]==leader1){
                count1++;
            }
            else if(nums[i]==leader2){
                count2++;
            }
        }
        if(count1>n/3){
            ans.push_back(leader1);
        }
        if(count2>n/3){
            ans.push_back(leader2);
        }
        return ans;
    }
};
int main(){
    int n;
    cout<<"Enter the size of the array: ";
    cin>>n;
    vector<int>nums(n);
    cout<<"Enter the elements of the array: ";
    for(int i=0;i<n;i++){
        cin>>nums[i];
    }
    MajorityElement2 obj;
    vector<int>result = obj.majorElement(nums);
    cout<<"The majority elements are: ";
    for(int i=0;i<result.size();i++){
        cout<<result[i]<<" ";
    }
    return 0;
}